### PR TITLE
Service binding resolver

### LINKED ISSUE #107: Support retrieving bindings by buildpacks
#### Context
Some buildpacks may accept settings/credentials using [bindings](https://github.com/k8s-service-bindings/spec) at build or runtime. A common example would be the host location and credential for an external service.

#### Issue request
Packit should provide a way for a buildpack to look up and retrieve these bindings.

#### Example Use case.

See the session handler buildpacks in this [proposed PHP RFC](https://github.com/paketo-buildpacks/php/blob/a1ef40f8b92906d6086c008124348fa953d38a1e/rfcs/0001-restructure.md). They expect their settings to be provided through a suitable binding.

#### Related links
https://paketo.io/docs/buildpacks/configuration/#bindings
https://github.com/paketo-buildpacks/rfcs/blob/main/accepted/0010-dependency-mappings.md
https://github.com/paketo-buildpacks/maven/tree/v3.2.1#bindings

### PR DESCRIPTION
**Pull request by menehune23**
## Summary
Adds a `servicebindings.Resolver` type which can be used to lazily load and resolve service bindings according to the [k8s spec](https://github.com/k8s-service-bindings/spec#workload-projection).

Implementation notes:

- Because the new spec relies on a `$SERVICE_BINDING_ROOT` env var, this PR maintains backwards compatibility by falling back to `$CNB_BINDINGS`, then `<platform>/bindings` if neither env var is set.

- Bindings can be read from either the new [k8s format](https://github.com/k8s-service-bindings/spec#workload-projection) or the [legacy format](https://github.com/buildpacks/spec/blob/main/extensions/bindings.md).

- Rather than providing top-level functions, this PR uses a dedicated type so that consumers can create interfaces for it and also mock/fake it in unit tests to keep them simple (for instance, this eliminates the need to understand the spec's implementation and write files to the filesystem during unit tests that rely on a binding).

- Entry values are of a new `io.ReadCloser`-compatible type, `Entry` -- rather than simply a `string` or `[]byte` -- that can stream the entry from the file system. It also supports `.ReadString()` and `.ReadBytes()` for ease of use. (Credit to @ryanmoran for the great idea).

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

Closes #107

### CODE REVIEW

File: bindings/resolver.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	// Type is the type of the binding, given by the content of the 'type' file within the binding directory.
+	Type string
+
+	// Provider is the provider of the binding, given by the content of the 'provider' file within the binding
+	// directory.
+	Provider string
+
+	// Secret is the primary content of the binding. Keys are given by each file name within the binding directory
+	// (other than 'type' or 'provider'), and corresponding values are given by the content of each file.
+	Secret map[string][]byte // TODO: place in custom buffer class
```
Comment (menehune23) [inline bindings/resolver.go:30]: Note the TODO: the rhs of this map will implement the `io.Reader` interface

File: postal/internal/dependency_mappings_test.go [hunk 2]
```diff
@@ -66,45 +66,7 @@ func testDependencyMappings(t *testing.T, context spec.G, it spec.S) {
...
-		context("when SHA256 contents cannot be opened", func() {
-			it.Before(func() {
-				Expect(os.MkdirAll(filepath.Join(bindingPath, "some-binding"), 0700)).To(Succeed())
-				Expect(os.WriteFile(filepath.Join(bindingPath, "some-binding", "type"), []byte("dependency-mapping"), 0600)).To(Succeed())
-				Expect(os.WriteFile(filepath.Join(bindingPath, "some-binding", "some-sha"), []byte("dependency-mapping-entry.tgz"), 0000)).To(Succeed())
-			})
-			it("errors", func() {
-				_, err := resolver.FindDependencyMapping("some-sha", bindingPath)
-				Expect(err).To(HaveOccurred())
+				Expect(err).To(MatchError(ContainSubstring("failed to list service bindings")))
```
Comment (menehune23) [inline postal/internal/dependency_mappings_test.go:69]: Haven't updated this file yet...

File: postal/service.go [hunk 3]
```diff
@@ -2,6 +2,8 @@ package postal
 
 import (
 	"fmt"
+	"github.com/paketo-buildpacks/packit/bindings"
```
Comment (menehune23) [inline postal/service.go:5]: Will fix ordering

File: bindings/resolver.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+	delete(metadata, "provider")
+
+	binding.Secret = metadata
+
+	secrets, err := loadSecrets(filepath.Join(binding.Path, "secret"))
+	if err != nil && !os.IsNotExist(err) {
+		return Binding{}, err
+	}
+	if err == nil {
+		for k, v := range secrets {
```
Comment (menehune23) [inline bindings/resolver.go:230]: Technically, this means that any keys within `secret` will trump keys within `metadata`. I don't love it, but it's how existing buildpack implementations handle legacy bindings (via [libcnb](https://github.com/buildpacks/libcnb/blob/main/platform.go#L100)), so it's nothing new.

To handle it differently, there are a couple of options:
- Return two different binding types when resolving: `Binding` or `LegacyBinding` (meh)
- Open up `Binding` to separate `metadata` from `secret` into separate fields (results in unused field for newer binding type and could be confusing)

Overall, it's such an edge case that I don't think it's worth those refactors to address, as it's only for the legacy binding format. No ambiguity for the new format.

Thoughts?

File: postal/internal/dependency_mappings_test.go [hunk 5]
```diff
@@ -32,80 +33,48 @@ func testDependencyMappings(t *testing.T, context spec.G, it spec.S) {
...
-			Expect(os.WriteFile(filepath.Join(bindingPath, "some-binding", "some-sha"), []byte("dependency-mapping-entry.tgz"), 0600)).To(Succeed())
-
-			Expect(os.MkdirAll(filepath.Join(bindingPath, "other-binding"), 0700)).To(Succeed())
-			Expect(os.WriteFile(filepath.Join(bindingPath, "other-binding", "type"), []byte("dependency-mapping"), 0600)).To(Succeed())
-			Expect(os.WriteFile(filepath.Join(bindingPath, "other-binding", "other-sha"), []byte("dependency-mapping-entry.tgz"), 0600)).To(Succeed())
-
-			Expect(os.MkdirAll(filepath.Join(bindingPath, "another-binding"), 0700)).To(Succeed())
-			Expect(os.WriteFile(filepath.Join(bindingPath, "another-binding", "type"), []byte("another type"), 0600)).To(Succeed())
-			Expect(os.WriteFile(filepath.Join(bindingPath, "another-binding", "some-sha"), []byte("entry.tgz"), 0600)).To(Succeed())
+			bindingResolver.ResolveCall.Returns.BindingSlice = []bindings.Binding{
```
Comment (menehune23) [inline postal/internal/dependency_mappings_test.go:36]: No need to write to the file system for this unit test anymore! 🎉

File: bindings/resolver.go [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// SecretReader represents the read-only content of a secret value.
```
Comment (menehune23) [inline bindings/resolver.go:14]: Technically this isn't specific to secrets/bindings. Should we move this type out to a more general location in `packit` for use in other places? (renaming to something like `ByteReader` or similar)

File: postal/internal/fakes/binding_resolver.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
+package fakes
```
Comment (menehune23) [inline postal/internal/fakes/binding_resolver.go:1]: Generated via `faux`

File: postal/fakes/mapping_resolver.go [hunk 8]
```diff
@@ -7,25 +7,23 @@ type MappingResolver struct {
 		sync.Mutex
 		CallCount int
 		Receives  struct {
-			SHA256      string
-			BindingPath string
+			SHA256 string
```
Comment (menehune23) [inline postal/fakes/mapping_resolver.go:10]: Generated via `faux`

File: bindings/resolver.go [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	for _, file := range files {
+		content, err := os.ReadFile(filepath.Join(secretsPath, file.Name()))
+		if err != nil {
+			return nil, err
+		}
+		secrets[file.Name()] = MakeSecretReader(content)
```
Comment (ryanmoran) [inline bindings/resolver.go:250]: I think I've come around to the idea that we should not read the whole contents into memory before the point at which someone calls `String()`, `Bytes()`, or `Read(b []byte)` on the `SecretReader`. Instead, could the implementation lazily read these contents when those methods are invoked? I think we'd also need to add `Close()` to the API contract to make this work well without exhausting file descriptors in some edgecases.

Basically, I am concerned about a case where a secret is a very large file, like a binary or tarball, that we then read the entire contents of into memory.

File: bindings/resolver.go (hunk 4 above)
Comment (sophiewigmore) [reply bindings/resolver.go:230]: I'm not sure how often the collision will occur, but I'm open to the second option you've proposed.  I would rather that the case for the k8s-based binding spec was adhered more strictly (no empty fields) than the legacy-based binding  since that's the direction we are moving in in the long run. 

I would also be okay with leaving the code as is in this case, and then revisiting this if we get user feedback about this being an issue.

File: bindings/resolver.go (hunk 6 above)
Comment (menehune23) [reply bindings/resolver.go:14]: Sounds good 👍

File: bindings/resolver.go [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+	// Type is the type of the binding, given by the content of the 'type' file within the binding directory.
+	Type string
+
+	// Provider is the provider of the binding, given by the content of the 'provider' file within the binding
+	// directory.
+	Provider string
+
+	// Secret is the primary content of the binding. Keys are given by each file name within the binding directory
+	// (other than 'type' or 'provider'), and corresponding values are given by the content of each file.
+	Secret map[string]SecretReader
```
[the remark this answers is not in the document]
Comment (menehune23) [reply bindings/resolver.go:60]: Ah, good question. This is mostly a carryover from the terminology used by the [old spec](https://github.com/buildpacks/spec/blob/main/extensions/bindings.md).

Now that I think about it though, I think even "secret/secrets" is too specific. The newer spec calls them "entries". What do you think of renaming this to `Entries`?

File: bindings/resolver.go [hunk 11]
```diff
@@ -0,0 +1,253 @@
+package bindings
```
[the remark this answers is not in the document]
Comment (menehune23) [reply bindings/resolver.go:1]: Yeah, I ran into that myself (and used `binds` for the variable to resolve it). I considered `servicebindings` as a package name, but it felt long. Happy to change to that or take other suggestions.

File: bindings/resolver.go [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// SecretReader represents the read-only content of a secret value.
+type SecretReader struct {
+	data *bytes.Buffer
+}
```
[the remark this answers is not in the document]
Comment (menehune23) [reply bindings/resolver.go:17]: No particular reason other than it went along with the `Binding` struct and was small. I'll move it out :)

File: bindings/resolver.go [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// SecretReader represents the read-only content of a secret value.
+type SecretReader struct {
```
[the remark this answers is not in the document]
Comment (menehune23) [reply bindings/resolver.go:15]: I suppose I chose it because it felt most descriptive of its primary purpose. I feel you though and am not super attached to the name.

Given the other thread about ranaming `Binding.Secret` to `Binding.Entries`, perhaps `Entry` would make sense name here?

File: bindings/resolver.go [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+			resolved = append(resolved, bind)
+		}
+	}
+	return resolved, nil
+}
+
+// ResolveOne returns a single binding matching the given type and optional provider (case-insensitive). To match on
+// type only, provider may be an empty string. Returns an error if the number of matched bindings is not exactly one, or
+// if there are problems loading bindings from the file system.
+func (r *Resolver) ResolveOne(typ string, provider string) (Binding, error) {
```
[the remark this answers is not in the document]
Comment (menehune23) [reply bindings/resolver.go:115]: @sophiewigmore I debated this too. I kept it because it's implemented similarly in `libpak`, and I also saw at least one buildpack implementation use it (can't remember off-hand). So I felt like it might be good as an option. I was also planning to use it in the dependency mapping logic here in `packit` but it turned out not to need it.

[the remark this answers is not in the document]
Comment (menehune23) [reply bindings/resolver.go:115]: I could see it being useful when a buildpack expects only one binding for a given type and wouldn't be able to decide what to do if it were ambiguous anyway.

File: bindings/resolver.go (hunk 9 above)
Comment (menehune23) [reply bindings/resolver.go:250]: That shouldn't be difficult to change. I'll look into it. I'll have to move away from using `bytes.Buffer` under the hood and may just use the file object

File: bindings/resolver.go (hunk 4 above)
Comment (menehune23) [reply bindings/resolver.go:230]: OK, I'm also in favor of leaving as is. I think if this were a big risk, it would have turned up by now with all the Java buildpacks' implementations (which use `libcnb` under the hood).

File: bindings/resolver.go [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+	}
+
+	return &Resolver{
+		bindingRoot: root,
+	}
+}
+
+// Resolve returns all bindings matching the given type and optional provider (case-insensitive). To match on type only,
+// provider may be an empty string. Returns an error if there are problems loading bindings from the file system.
+func (r *Resolver) Resolve(typ string, provider string) ([]Binding, error) {
```
[the remark this answers is not in the document]
Comment (menehune23) [reply bindings/resolver.go:93]: I know that it's the convention for interface naming (e.g. `Stringer.String()`), but I understand that it may not be as desirable for structs. I just don't know why and have no personal reason for avoiding it (other than sometimes name collisions with interfaces). It also seems hard to avoid when trying to come up with descriptive names.

I'd personally prefer to leave as is, if it's just a nit.

But if you feel more strongly, I'm not opposed to renaming. I think that the names above are perhaps ambiguous as to the plurality (mostly on `Get` as it doesn't necessarily imply one), so if I were to change this, I might suggest something like `Get` and `GetOne`, or `Filter` and `FilterOne`, or something like that.

File: servicebindings/entry.go [hunk 16]
```diff
@@ -0,0 +1,31 @@
+package servicebindings
+
+import "bytes"
+
+// Entry represents the read-only content of a binding entry.
+type Entry struct {
```
Comment (menehune23) [inline servicebindings/entry.go:6]: @ryanmoran I still need to make this adhere to your lazy-read suggestion

File: bindings/resolver.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (menehune23) [reply bindings/resolver.go]: Done.

File: bindings/resolver.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (menehune23) [reply bindings/resolver.go]: Renamed to `servicebindings`

File: bindings/resolver.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (menehune23) [reply bindings/resolver.go]: Done.

File: servicebindings/entry.go (hunk 16 above)
Comment (menehune23) [reply servicebindings/entry.go:8]: Done. Interested in your thoughts on the newest commit @ryanmoran

File: bindings/resolver.go (hunk 9 above)
Comment (menehune23) [reply bindings/resolver.go]: Done

File: postal/internal/dependency_mappings.go [hunk 17]
```diff
@@ -1,52 +1,38 @@
...
 
-// Given a target dependency, look up if there is a matching dependency mapping at the given binding path
-func (d DependencyMappingResolver) FindDependencyMapping(sha256, bindingPath string) (string, error) {
-	allBindings, err := filepath.Glob(filepath.Join(bindingPath, "*"))
+// FindDependencyMapping looks up if there is a matching dependency mapping at the given binding path
+func (d DependencyMappingResolver) FindDependencyMapping(sha256 string) (string, error) {
+	bindings, err := d.bindingResolver.Resolve("dependency-mapping", "")
 	if err != nil {
-		return "", err
+		return "", errors.Wrap(err, "getting dependency-mapping binding")
```
Comment (ryanmoran) [inline postal/internal/dependency_mappings.go:28]: Rather than pulling in a whole new package, maybe we can use the stdlib built-in error wrapping?

```suggestion
		return "", fmt.Errorf("failed to resolve dependency-mapping binding: %w", err)
```

File: postal/service.go [hunk 18]
```diff
@@ -39,10 +41,12 @@ type Service struct {
 }
 
 // NewService creates an instance of a Servicel given a Transport.
-func NewService(transport Transport) Service {
+func NewService(transport Transport, platformDir string) Service {
```
Comment (ryanmoran) [inline postal/service.go:44]: This changes the public API. Does this mean we want to bump the major version of `packit` as its a backwards-incompatible change?

Comment (menehune23) [reply postal/service.go:44]: Yes, that's correct. Do we have a way of recording this? (Could be a good case for something like https://github.com/semantic-release/semantic-release eventually)

File: postal/internal/dependency_mappings.go (hunk 17 above)
Comment (menehune23) [reply postal/internal/dependency_mappings.go]: Done here and elsewhere ✔️

File: postal/service.go (hunk 18 above)
Comment (ryanmoran) [reply postal/service.go:44]: I guess I bring this up because we *just* cut a `1.0`. I feel like we should be doing a better job of keeping some stability in the API for a while before we break things again. Is there a way we can make this backwards-compatible?

Comment (menehune23) [reply postal/service.go:44]: Ah, shoot. I see. Let me look and see if there's a way. The `MappingResolver` interface has changed too, so technically `WithDependencyMappingResolver` isn't backwards compatible either...

Comment (menehune23) [reply postal/service.go:44]: I wrote all this assuming a major version bump would result, but I didn't realize we just cut a `1.0` :(

Comment (ryanmoran) [reply postal/service.go:44]: The `MappingResolver` is in an internal package and therefore not part of the public API. I wouldn't worry about that change.

Comment (menehune23) [reply postal/service.go:44]: True that our mapping resolver implementation is internal, but `WithDependencyMappingResolver` and the `MappingResolver` interface are still public. That said, it would only break folks who are using their own implementation of a mapping resolver, as opposed to ours, so this probably isn't a big deal. In fact, given that, I'm not even sure any library consumer would ever use `WithDependencyMappingResolver` in the first place. I'm guessing we added for DI purposes for our own tests.

Comment (menehune23) [reply postal/service.go:44]: Here are our options for making this backwards-compatible:
- Hardcode the platform dir here so that we don't need to pass it in. That said, this doesn't actually seem feasible as it is passed in as an argument to build/detect.
- Since the platform dir needs to be fed in somewhere, it seems the only other option is to move it to a parameter on `Resolve`/`ResolveOne`. I could update the lazy-loading logic to store the last used platform dir and only re-read bindings from the filesystem if it has changed.

There may be some other more clever options if you have any other suggestions @ryanmoran

File: postal/service.go [hunk 19]
```diff
@@ -26,9 +28,9 @@ type Transport interface {
 
 //go:generate faux --interface MappingResolver --output fakes/mapping_resolver.go
 // MappingResolver serves as the interface that looks up platform binding provided
-// dependency mappings given a  SHA256 and a path to search for bindings
+// dependency mappings given a SHA256
 type MappingResolver interface {
-	FindDependencyMapping(SHA256, bindingPath string) (string, error)
+	FindDependencyMapping(SHA256 string, platformDir string) (string, error)
```
Comment (menehune23) [inline postal/service.go:33]: @ryanmoran This has become the platform dir rather than full binding path. Technically it's a change in the API, but as you stated earlier, the only real implementation of this interface is internal.

File: postal/service.go [hunk 20]
```diff
@@ -175,13 +179,6 @@ func (s Service) Deliver(dependency Dependency, cnbPath, layerPath, platformPath
 	return nil
 }
 
-// Install will invoke Deliver with a hardcoded value of /platform for the platform path.
-//
-// Deprecated: Use Deliver instead.
```
Comment (menehune23) [inline postal/service.go:180]: @ryanmoran Is this ok to deprecate now?

File: postal/service_test.go [hunk 21]
```diff
@@ -714,258 +716,6 @@ version = "this is super not semver"
 		})
 	})
 
-	context("Install", func() {
```
Comment (menehune23) [inline postal/service_test.go:717]: At a glance, these looked to be covered by the `"Deliver"` context tests. Feel free to correct me if I missed something.

File: servicebindings/resolver.go [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+// Resolve returns all bindings matching the given type and optional provider (case-insensitive). To match on type only,
+// provider may be an empty string. Returns an error if there are problems loading bindings from the file system.
+//
+// The location of bindings is given by one of the following, in order of precedence:
+//
+//   1. SERVICE_BINDING_ROOT environment variable
+//   2. CNB_BINDINGS environment variable, if above is not set
+//   3. `<platformDir>/bindings`, if both above are not set
+func (r *Resolver) Resolve(typ string, provider string, platformDir string) ([]Binding, error) {
+	if newRoot := bindingRoot(platformDir); r.bindingRoot != newRoot {
```
Comment (menehune23) [inline servicebindings/resolver.go:54]: Now that the platform is part of this method: If the binding root changes, we'll refresh and cache new bindings. Note that this implicitly occurs on a new resolver, due to the root being initialized to `""`.

File: postal/service.go (hunk 18 above)
Comment (menehune23) [reply postal/service.go]: @ryanmoran Created a new commit with the second option implemented. Let me know your thoughts. Should be easy to revert/adjust since I left it separate.

File: postal/service.go (hunk 20 above)
Comment (ryanmoran) [reply postal/service.go:180]: It is deprecated. Do you mean remove?

File: postal/service_test.go (hunk 21 above)
Comment (ryanmoran) [reply postal/service_test.go:717]: Yeah. This is fine. I always figured they just get removed when we removed the method.

File: servicebindings/resolver.go [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+
+// Resolve returns all bindings matching the given type and optional provider (case-insensitive). To match on type only,
+// provider may be an empty string. Returns an error if there are problems loading bindings from the file system.
+//
+// The location of bindings is given by one of the following, in order of precedence:
+//
+//   1. SERVICE_BINDING_ROOT environment variable
+//   2. CNB_BINDINGS environment variable, if above is not set
+//   3. `<platformDir>/bindings`, if both above are not set
+func (r *Resolver) Resolve(typ string, provider string, platformDir string) ([]Binding, error) {
```
Comment (ryanmoran) [inline servicebindings/resolver.go:53]: Did you know you can reduce some of the verbosity in these params by declaring types for sequential params that have that same type?

```suggestion
func (r *Resolver) Resolve(typ, provider, platformDir string) ([]Binding, error) {
```

**Comment by ryanmoran** [comment]
I'm happy with where this is heading now. @sophiewigmore do you have anything else you'd like to see before we think about getting this merged?

File: postal/service.go (hunk 20 above)
Comment (menehune23) [reply postal/service.go:180]: Oh yes, that's what I meant :)

Comment (menehune23) [reply postal/service.go:180]: Assuming you're good with this given your comments on the removal of associated tests.

File: servicebindings/resolver.go (hunk 23 above)
Comment (menehune23) [reply servicebindings/resolver.go:53]: Yeah, I know about that but have never preferred it (as it makes modification cumbersome IMO). Happy to conform with the patterns of this repo though.

Comment (menehune23) [reply servicebindings/resolver.go]: Done here and elsewhere

**Comment by menehune23** [comment]
> I'm happy with where this is heading now. @sophiewigmore do you have anything else you'd like to see before we think about getting this merged?

@ryanmoran @sophiewigmore Just let me know before you merge, and I'll clean up the commit history

**Comment by sophiewigmore** [comment]
@ryanmoran @menehune23 hey y'all. sorry I missed the update earlier. I've been following along with these discussions and this sounds good to me!

**Comment by ForestEckhardt** [comment]
@menehune23 I am still a little unsure about the removal of the `postal.Install` function. I know that it is deprecated but the removal of the function entirely feels like it would necessitate a major bump as we are changing a part of the API by removing functionality. If our goal is to not do a major bump then I feel like that would bee to be added back in unless I am missing something or the intention is to do a major bump and I have just missed the boat on the decision.

**Comment by menehune23** [comment]
> @menehune23 I am still a little unsure about the removal of the `postal.Install` function. I know that it is deprecated but the removal of the function entirely feels like it would necessitate a major bump as we are changing a part of the API by removing functionality. If our goal is to not do a major bump then I feel like that would bee to be added back in unless I am missing something or the intention is to do a major bump and I have just missed the boat on the decision.

@ForestEckhardt That's a good point. Perhaps what I'll do is restore the function but leave the tests removed (since the functionality is covered by `Deliver` tests).

**Review by ForestEckhardt — APPROVED** [review]

**Review by ForestEckhardt — APPROVED** [review]
